### PR TITLE
Add error handling to avoid problems with closed o bad fds

### PR DESCRIPTION
**Pull request by fortizc**
* Add function to remove and unsubscribe fds (logserver_remove_fd)
* Add error handling to catch bad fds on epoll
* Add error handling to remove bad on read functions
* Replace calls to close fds with logserver_remove_fd
* Fix unsubscription mechanism adding a new header to subscribe or unsubscribe

### CODE REVIEW

File: logserver.c [hunk 1]
```diff
@@ -498,7 +498,7 @@ static int logserver_handle_msg(struct logserver_msg *msg)
 static int logserver_epoll_command(int fd, int cmd)
 {
 	struct epoll_event ev;
-	ev.events = EPOLLIN;
+	ev.events = EPOLLIN | EPOLLRDHUP;
```
Comment (asac) [inline logserver.c:501]: we dont need EPOLLHUP here?

File: logserver.c [hunk 2]
```diff
@@ -678,16 +682,43 @@ static int logserver_epoll_wait(struct epoll_event *ev)
...
 	struct buffer *buffer = pv_buffer_get(true);
 
 	if (!buffer)
 		return;
 
+	errno = 0;
 	if (pv_fs_file_read_nointr(fd, buffer->buf, buffer->size) > 0) {
 		struct logserver_msg *msg = (struct logserver_msg *)buffer->buf;
 		logserver_handle_msg(msg);
+	} else if (errno != EAGAIN && errno != EINTR && errno != 0) {
```
Comment (asac) [inline logserver.c:718]: when is errno == 0 here?

File: logserver.c [hunk 3]
```diff
@@ -713,6 +745,10 @@ static void logserver_consume_fd(int fd)
 						.data = buffer->buf,
 						.data_len = size };
 		logserver_log_msg_data(&d);
+	} else if (errno != EAGAIN && errno != EINTR && errno != 0) {
```
Comment (asac) [inline logserver.c:748]: dito

File: logserver.c [hunk 4]
```diff
@@ -782,16 +812,27 @@ static void logserver_loop()
...
+			    curev & EPOLLERR) {
+				pv_log(WARN, "bad fd found (fd = %d)", curfd);
+				logserver_remove_fd(curfd);
+				continue;
+			}
+		}
+
 		if (curfd == logsock || curfd == fdsock) {
 			int fd = logserver_accept_connection(curfd);
 			if (fd < 0) {
```
Comment (asac) [inline logserver.c:830]: in which cases is accept_connection < 0 ? is that self healing?

File: logserver.c (hunk 1 above)
Comment (fortizc) [reply logserver.c:501]: No because EPOLLHUP is always reported by epoll_wait

**Pull request merged**